File: scripts/dependencies.js
const fs = require("fs");
const { dedent } = require("ts-dedent");

function getPackage(name) {
  const packageInfo = require(`../node_modules/${name}/package.json`);

  return {
    name: packageInfo.name,
    version: packageInfo.version,
    main: packageInfo.unpkg || packageInfo.main,
    style: packageInfo.style
  };
}

function generateDependencies() {
  const packages = [
    getPackage("mapbox-gl"),
    getPackage("h3-js"),
    getPackage("s2-geometry"),
    { ...getPackage("deck.gl"), main: "dist.min.js" },
    getPackage("d3-array"),
    getPackage("d3-interpolate"),
    getPackage("d3-color"),
    getPackage("d3-scale")
  ];

  const dependencies = packages.map(({ name, version, main, style }) =>
    dedent(`
      htmltools::htmlDependency(
        name = "${name}",
        version = "${version}",
        src = c(href = "https://unpkg.com/${name}@${version}"),
        script = "${main}",
        stylesheet = ${style ? `"${style}"` : "NULL"}
      )
    `)
  );

  const content = dedent(`
    dependencies <- list(
      ${dependencies.join(",\n")}
    )
  `);

  fs.writeFileSync("./R/dependencies.R", content);
}

module.exports = {
  generateDependencies
};

generateDependencies();

File: R/dependencies.R
dependencies <- list(
  htmltools::htmlDependency(
    name = "mapbox-gl",
    version = "1.8.1",
    src = c(href = "https://unpkg.com/mapbox-gl@1.8.1"),
    script = "dist/mapbox-gl.js",
    stylesheet = "dist/mapbox-gl.css"
  ),
  htmltools::htmlDependency(
    name = "h3-js",
    version = "3.6.3",
    src = c(href = "https://unpkg.com/h3-js@3.6.3"),
    script = "dist/h3-js.umd.js",
    stylesheet = NULL
  ),
  htmltools::htmlDependency(
    name = "s2-geometry",
    version = "1.2.10",
    src = c(href = "https://unpkg.com/s2-geometry@1.2.10"),
    script = "src/s2geometry.js",
    stylesheet = NULL
  ),
  htmltools::htmlDependency(
    name = "deck.gl",
    version = "8.1.0",
    src = c(href = "https://unpkg.com/deck.gl@8.1.0"),
    script = "dist.min.js",
    stylesheet = NULL
  ),
  htmltools::htmlDependency(
    name = "d3-array",
    version = "2.4.0",
    src = c(href = "https://unpkg.com/d3-array@2.4.0"),
    script = "dist/d3-array.min.js",
    stylesheet = NULL
  ),
  htmltools::htmlDependency(
    name = "d3-interpolate",
    version = "1.4.0",
    src = c(href = "https://unpkg.com/d3-interpolate@1.4.0"),
    script = "dist/d3-interpolate.min.js",
    stylesheet = NULL
  ),
  htmltools::htmlDependency(
    name = "d3-color",
    version = "1.4.0",
    src = c(href = "https://unpkg.com/d3-color@1.4.0"),
    script = "dist/d3-color.min.js",
    stylesheet = NULL
  ),
  htmltools::htmlDependency(
    name = "d3-scale",
    version = "3.2.1",
    src = c(href = "https://unpkg.com/d3-scale@3.2.1"),
    script = "dist/d3-scale.min.js",
    stylesheet = NULL
  )
)
